### PR TITLE
Training deep SNNs with DECOLLE

### PR DESCRIPTION
**Pull request by nskat**
Issue Number:
 #97 
 
Objective of pull request:
Implementation of the training algorithm DECOLLE (Kaiser et al. 2020)

## Pull request checklist
Your PR fulfills the following requirements:
- [x] [Issue](https://github.com/lava-nc/lava-dl/issues) created that explains the change and why it's needed
- [ ] Tests are part of the PR (for bug fixes / features)
- [ ] [Docs](https://github.com/lava-nc/docs) reviewed and added / updated if needed (for bug fixes / features)
- [x] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
- [x] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
- [x] Lint (`flakeheaven lint src/lava tests/`) and (`bandit -r src/lava/.`) pass locally
- [x] Build tests (`pytest`) passes locally

## Pull request type
Please check your PR type:
- [ ] Bugfix
- [x ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation changes
- [ ] Other (please describe): 

## What is the current behavior?
Currently, only training with SLAYER is implemented and documented. 
As per the desire of the developers to provide a general framework for neuromorphic computing, I believe that it would be great if Lava implemented training of deep SNNs via a variety of algorithms. In that sense, the naming of the training library (lava.lib.dl.slayer) would also benefit from evolving in future iterations, e.g. as lava.lib.dl.training. 

## What is the new behavior?
I have implemented a working example of training an SNN with DECOLLE in the tutorials with an assorted jupyter notebook tutorial. I have also added a custom loss in dl.slayer.loss and a custom training assistant in dl.slayer.utils.assistant.

## Does this introduce a breaking change?
- [ ] Yes
- [x] No

## Supplemental information
Please let me know if I can improve my PR in any way.
In particular, the code closely follows that used for training with SLAYER, so I didn't feel the need to add many comments but I'm happy to add them wherever they lack. Also, I did not add unit tests as there were none covering the counterparts to what I have implemented in SLAYER but I can add some if you feel they are required.

### CODE REVIEW

**Comment by bamsumit** [comment]
@nskat Thanks for the PR contribution. DECOLLE is a valuable addition to lava-dl. We will review it soon and provide feedback soon.

At first glance, can you remove the files inside `.idea/`. They look unnecessary.

**Comment by nskat** [comment]
Hi @bamsumit, thank you for your response, I've deleted the files. Sorry about that, they were added by Pycharm.

File: src/lava/lib/dl/slayer/loss.py [hunk 1]
```diff
@@ -210,3 +210,43 @@ def forward(self, input, label):
...
+
+class DECOLLELoss(object):
+    """
+    Adapted from https://github.com/nmi-lab/decolle-public
+    Computes the DECOLLE Loss for the model, defined as the sum of the per-layer
+    local pseudo-losses + a regularization term
+    """
+
+    def __init__(self, loss_fn, net, reg=0, reduction='mean'):
+        """'
```
Comment (bamsumit) [inline src/lava/lib/dl/slayer/loss.py:223]: Let's follow numpy docstring format and make the description formal. Also, we follow the convention of writing docstring for the constructor with the class docstring.

File: src/lava/lib/dl/slayer/utils/assistant.py [hunk 2]
```diff
@@ -258,3 +258,287 @@ def valid(self, input, target):
...
+    ----------
+    net
+    error
+    optimizer
+    stats
+    classifier
+    count_log
+    lam
+    device : torch.device or None
+        the main device memory where network is placed. It is not at start and
```
Comment (bamsumit) [inline src/lava/lib/dl/slayer/utils/assistant.py:304]: It is None at start?

File: tutorials/lava/lib/dl/slayer/nmnist/train.ipynb [hunk 3]
```diff
@@ -3,7 +3,11 @@
   {
```
Comment (bamsumit) [inline tutorials/lava/lib/dl/slayer/nmnist/train.ipynb:3]: I don't think there are any changes except your local run metadata. So this can perhaps be removed from the PR

**Review by bamsumit — CHANGES_REQUESTED** [review]
The implementation looks good. I would suggest changing the directory structure like this
```
src/lava/lib/dl
              |-slayer
              |-bootstrap
              |-decolle
              |       |-README.md
              |       |-loss.py
              |       |-utils
              |             |- __init__.py 
              |             |- assistant.py 
              |- netx 
tutorials/lava/lib/dl
                    |-slayer
                    |   |-nmnist
                    |   |         |-nmnist.py
                    |   |         |-train.ipynb
                    |   |-oxford
                    |   |-...
                    |-decolle
                            |-nmnist
                                   |-nmnist.py (nmnist_decolle.py)
                                   |-train.ipynb (train_decolle.py)
```
**Decolle** would then be a training method in **lava-dl** similar to **slayer** and **bootstrap**.

You will just need to move around your current implementation.

Also, add a `README.md` with a brief explanation of what DECOLLE does. __Do not forget to mention the paper and add a link the original DECOLLE github implementation.__

I can take care of updating the lava-dl graphic with decolle.

File: src/lava/lib/dl/slayer/loss.py (hunk 1 above)
Comment (nskat) [reply src/lava/lib/dl/slayer/loss.py]: Thank you, I have updated the docstring now

File: src/lava/lib/dl/slayer/utils/assistant.py (hunk 2 above)
Comment (nskat) [reply src/lava/lib/dl/slayer/utils/assistant.py]: It is, it was a typo in the Assistant docstring, I have fixed both

File: tutorials/lava/lib/dl/slayer/nmnist/train.ipynb (hunk 3 above)
Comment (nskat) [reply tutorials/lava/lib/dl/slayer/nmnist/train.ipynb]: My bad, I have deleted this

**Comment by nskat** [comment]
Hi @bamsumit, thanks a lot for your review, I have made the requested changes. I'll let you update the lava-dl graphic with decolle.

File: tutorials/lava/lib/dl/decolle/README.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# DECOLLE
```
Comment (bamsumit) [inline tutorials/lava/lib/dl/decolle/README.md:1]: @nskat Readme looks good.
Let's move it to `src/lava/lib/dl/decolle`

File: src/lava/lib/dl/slayer/utils/assistant.py [hunk 5]
```diff
@@ -258,3 +258,287 @@ def valid(self, input, target):
                 return output
 
             return output, count
+
+
+class DECOLLEAssistant(Assistant):
```
Comment (bamsumit) [inline src/lava/lib/dl/slayer/utils/assistant.py:263]: Let's move this class to `src/lava/lib/dl/decolle/utils/assistant.py`. Please also add `__init__.py`.

File: src/lava/lib/dl/slayer/loss.py [hunk 6]
```diff
@@ -210,3 +210,54 @@ def forward(self, input, label):
                 label.flatten(),
                 reduction=self.reduction,
             )
+
+
+class DECOLLELoss(object):
```
Comment (bamsumit) [inline src/lava/lib/dl/slayer/loss.py:215]: Let's move this class definition to `src/lava/lib/dl/decolle/loss.py`. Also remember to add `__init__.py` file to the folder.

**Review by bamsumit — COMMENTED** [review]
Hi @nskat, your tutorial reorganization looks good. I also meant moving the decolle implementation in `src/lava/lib/dl/decolle` as well in my earlier comment, i.e.
* move `README.md` to `src/lava/lib/dl/decolle`
* move `DECOLLELoss` to `src/lava/lib/dl/decolle/loss.py`
* `__init__.py` in `src/lava/lib/dl/decolle` to expose `decolle.loss` and `decolle.utils`.
* move `DECOLLEAssistant` tp `src/lava/lib/dl/decolle/utils/assistant.py`
* `__init__.py` in `src/lava/lib/dl/decolle/utils` to expose `decolle.utils.assistant`.

**Review by tim-shea — APPROVED** [review]
Thank you for your contribution! Apart from Sumit's suggested reorganization, I only have two significant comments.

1) Emre Neftci's lab has released DECOLLE under GPL, so Lava should not change that license without permission. Unless you (@nskat) are affiliated with NMI-Lab @ UC Irvine or otherwise have approval to do so, the easiest would be to just GPL all the new code. To make that easy, we definitely should adopt Sumit's code structure in which all DECOLLE algos are in separate modules.

2) This is more a suggestion for the future, but I'd love to see a comparison between DECOLLE and SLAYER performance for some benchmark task. Such a thing could make a useful tutorial eventually.

**Comment by bamsumit** [comment]
Hi @nskat, thanks for making the changes. It looks good. 
__With the current state of `DECOLLELoss` implementation, we have some licensing concerns. Will you be able to reimplement it or derive the class from `decolle.base_model.DECOLLELoss` in the main decolle repo?__ That would simplify the licensing issue.

**Comment by nskat** [comment]
Hi @bamsumit, I have updated the implementation of the loss to inherit from Pytorch losses, which I think is in the spirit of lava-dl's implementation, and protects us from changes in the original decolle repo, which seems to be still updated sometimes.
Please let me know if this fixes the licensing issues -- I'm not familiar with these problems.

**Comment by bamsumit** [comment]
Hi @nskat can you contact me at sumit.bam.shrestha(at)intel.com ?
